File: src-tauri/src/commands/whatsapp.rs
use tauri::{command, AppHandle, State};
use crate::state::AppState;
use serde_json::Value;
use std::sync::Arc;
use crate::errors::AppError;

#[command]
pub async fn cmd_wa_start_session(
    app: AppHandle,
    state: State<'_, Arc<AppState>>,
    session_id: String,
) -> Result<Value, AppError> {
    let info = state.wa_manager.start_session(&app, &session_id).await?;
    Ok(serde_json::to_value(info).unwrap())
}

#[command]
pub async fn cmd_wa_stop_session(
    state: State<'_, Arc<AppState>>,
    session_id: String,
) -> Result<(), AppError> {
    state.wa_manager.stop_session(&session_id).await
}

#[command]
pub async fn cmd_wa_delete_session(
    app: AppHandle,
    state: State<'_, Arc<AppState>>,
    session_id: String,
) -> Result<(), AppError> {
    state.wa_manager.delete_session(&app, &session_id).await
}

#[command]
pub async fn cmd_wa_list_sessions(
    app: AppHandle,
    state: State<'_, Arc<AppState>>,
) -> Result<Value, AppError> {
    let sessions = state.wa_manager.list_sessions(&app).await;
    Ok(serde_json::to_value(sessions).unwrap())
}

#[command]
pub async fn cmd_wa_get_status(
    state: State<'_, Arc<AppState>>,
    session_id: String,
) -> Result<Value, AppError> {
    state.wa_manager.send_request(&session_id, "GET", "/status", None).await
}

#[command]
pub async fn cmd_wa_get_qr_url(
    state: State<'_, Arc<AppState>>,
    session_id: String,
) -> Result<String, AppError> {
    let port = state.wa_manager.get_session_port(&session_id).await?;
    Ok(format!("http://127.0.0.1:{}/qr", port))
}

#[command]
pub async fn cmd_wa_proxy_request(
    state: State<'_, Arc<AppState>>,
    session_id: String,
    method: String,
    path: String,
    body: Option<Value>,
) -> Result<Value, AppError> {
    state.wa_manager.send_request(&session_id, &method, &path, body.as_ref()).await
}

File: packages/api/src/modules/whatsapp.ts
import { invoke } from '@tauri-apps/api/core';
import type { WhatsAppSessionInfo } from '../types';

export const whatsapp = {
  startSession: (sessionId: string) => invoke<WhatsAppSessionInfo>('cmd_wa_start_session', { sessionId }),
  stopSession: (sessionId: string) => invoke<void>('cmd_wa_stop_session', { sessionId }),
  deleteSession: (sessionId: string) => invoke<void>('cmd_wa_delete_session', { sessionId }),
  listSessions: () => invoke<WhatsAppSessionInfo[]>('cmd_wa_list_sessions'),
  getStatus: (sessionId: string) => invoke<{ connected: boolean; jid: string }>('cmd_wa_get_status', { sessionId }),
  getQrUrl: (sessionId: string) => invoke<string>('cmd_wa_get_qr_url', { sessionId }),
  proxyRequest: <T = unknown>(sessionId: string, method: string, path: string, body?: unknown) =>
    invoke<T>('cmd_wa_proxy_request', { sessionId, method, path, body: body ?? null }),
};
